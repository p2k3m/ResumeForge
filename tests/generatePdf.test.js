import { jest } from '@jest/globals';
import { generatePdf, parseContent } from '../server.js';
import puppeteer from 'puppeteer';
import pdfParse from 'pdf-parse/lib/pdf-parse.js';
import zlib from 'zlib';

describe('generatePdf and parsing', () => {
  test('parseContent handles markdown', () => {
    const data = parseContent('Jane Doe\n# Education\n- Item 1\nText');
    const education = data.sections.find((s) => s.heading === 'Education');
    expect(education).toBeDefined();
    expect(education.items[0].map((t) => t.text).join('')).toBe('Item 1');
    expect(education.items[1].map((t) => t.text).join('')).toBe('Text');
  });

  test('parseContent handles JSON structure', () => {
    const json = { name: 'John', sections: [{ heading: 'Skills', items: ['JS'] }] };
    const data = parseContent(JSON.stringify(json));
    expect(data.name).toBe('John');
    const skills = data.sections.find((s) => s.heading === 'Skills');
    expect(skills.items[0].map((t) => t.text).join('')).toBe('JS');
  });

  test('parseContent creates multiple sections from headings', () => {
    const input = 'Jane Doe\n# Experience\n- Worked\n# Skills\n- JavaScript';
    const data = parseContent(input);
    expect(data.sections.map((s) => s.heading)).toEqual(['Experience', 'Skills']);
  });

  test.each(['modern', 'ucmo', 'professional', 'vibrant', '2025'])('generatePdf creates PDF from %s template', async (tpl) => {
    const buffer = await generatePdf('Jane Doe\n- Loves testing', tpl);
    expect(Buffer.isBuffer(buffer)).toBe(true);
    expect(buffer.length).toBeGreaterThan(0);
  });

  test('bold and italic text render without markdown syntax', () => {
    const data = parseContent(
      'Jane Doe\n- This is **bold** text\n- This is _italic_ text'
    );
    const [boldTokens, italicTokens] = data.sections[0].items;
    expect(boldTokens.map((t) => t.text).join('')).toBe('This is bold text');
    expect(italicTokens.map((t) => t.text).join('')).toBe('This is italic text');
    expect(boldTokens).toEqual(
      expect.arrayContaining([expect.objectContaining({ text: 'bold', style: 'bold' })])
    );
    expect(italicTokens).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ text: 'italic', style: 'italic' })
      ])
    );
    boldTokens.forEach((t) => expect(t.text).not.toContain('**'));
    italicTokens.forEach((t) => expect(t.text).not.toContain('_'));
  });

  test('spacing is preserved across multiple tokens', () => {
    const data = parseContent(
      'Jane Doe\n- Visit [OpenAI](https://openai.com) and [GitHub](https://github.com)\n- Mix **bold** and _italic_ styles'
    );
    const [linkTokens, mixTokens] = data.sections[0].items;
    expect(linkTokens.map((t) => t.text).join('')).toBe(
      'Visit OpenAI and GitHub'
    );
    expect(mixTokens.map((t) => t.text).join('')).toBe(
      'Mix bold and italic styles'
    );
  });

  test('line breaks and tabs retained for mixed bullet and paragraph content', () => {
    const input = 'Jane Doe\n- First bullet\n\tContinuation paragraph\nFinal line';
    const data = parseContent(input);
    const [first, second] = data.sections[0].items;
    expect(first).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ text: 'First bullet' }),
        expect.objectContaining({ type: 'newline' }),
        expect.objectContaining({ type: 'tab' }),
        expect.objectContaining({ text: 'Continuation paragraph' })
      ])
    );
    const rendered = first
      .map((t) => {
        if (t.type === 'newline') return '<br>';
        if (t.type === 'tab') return '<span class="tab"></span>';
        return t.text;
      })
      .join('');
    expect(rendered).toBe(
      'First bullet<br><span class="tab"></span>Continuation paragraph'
    );
    expect(second.map((t) => t.text).join('')).toBe('Final line');
  });

  test('multi-line bullet spacing maintained in JSON input', () => {
    const json = {
      name: 'Jane Doe',
      sections: [{ heading: 'Work', items: ['First line\n\tSecond line'] }]
    };
    const data = parseContent(JSON.stringify(json));
    const [tokens] = data.sections[0].items;
    expect(tokens).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ text: 'First line' }),
        expect.objectContaining({ type: 'newline' }),
        expect.objectContaining({ type: 'tab' }),
        expect.objectContaining({ text: 'Second line' })
      ])
    );
    const rendered = tokens
      .map((t) => {
        if (t.type === 'newline') return '<br>';
        if (t.type === 'tab') return '<span class="tab"></span>';
        return t.text;
      })
      .join('');
    expect(rendered).toBe('First line<br><span class="tab"></span>Second line');
  });

  test('single asterisk italic and bullet handling', () => {
    const data = parseContent(
      'Jane Doe\n* This has *italic* text'
    );
    const [tokens] = data.sections[0].items;
    expect(tokens.map((t) => t.text).join('')).toBe(
      'This has italic text'
    );
    expect(tokens).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ text: 'italic', style: 'italic' })
      ])
    );
    tokens.forEach((t) => expect(t.text).not.toContain('*'));
  });

  test('parseContent detects links', () => {
    const data = parseContent(
      'Jane Doe\n- Check [OpenAI](https://openai.com) and https://example.com'
    );
    const tokens = data.sections[0].items[0];
    expect(tokens).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          type: 'link',
          text: 'OpenAI',
          href: 'https://openai.com'
        }),
        expect.objectContaining({
          type: 'link',
          text: 'https://example.com',
          href: 'https://example.com'
        })
      ])
    );
  });

  test('generated PDF contains clickable links without HTML anchors', async () => {
    jest.spyOn(puppeteer, 'launch').mockRejectedValue(new Error('no browser'));
    const input =
      'John Doe\n- https://www.linkedin.com/in/johndoe\n- https://github.com/johndoe';
    const buffer = await generatePdf(input);
    const items = parseContent(input).sections[0].items;
    expect(items).toEqual(
      expect.arrayContaining([
        expect.arrayContaining([
          expect.objectContaining({
            type: 'link',
            text: 'LinkedIn',
            href: 'https://www.linkedin.com/in/johndoe'
          })
        ]),
        expect.arrayContaining([
          expect.objectContaining({
            type: 'link',
            text: 'GitHub',
            href: 'https://github.com/johndoe'
          })
        ])
      ])
    );
    const raw = buffer.toString();
    expect(raw).toContain('https://www.linkedin.com/in/johndoe');
    expect(raw).toContain('https://github.com/johndoe');
    expect(raw).not.toMatch(/<a[\s>]/);
  });

  test('PDFKit fallback matches Puppeteer output text', async () => {
    const input = 'Jane Doe\n- Bullet point';
    const browserPdf = await generatePdf(input, 'modern');
    jest.spyOn(puppeteer, 'launch').mockRejectedValue(new Error('no browser'));
    const fallbackPdf = await generatePdf(input, 'modern');
    try {
      const browserText = (await pdfParse(browserPdf)).text.trim();
      const fallbackText = (await pdfParse(fallbackPdf)).text.trim();
      expect(browserText).toContain('â€¢ Bullet point');
      expect(fallbackText).toBe(browserText);
    } catch {
      const extract = (pdf) => {
        let idx = 0;
        let text = '';
        while ((idx = pdf.indexOf(Buffer.from('stream'), idx)) !== -1) {
          const nl = pdf.indexOf('\n', idx) + 1;
          const end = pdf.indexOf(Buffer.from('endstream'), nl);
          const chunk = pdf.slice(nl, end);
          let content;
          try {
            content = zlib.inflateSync(chunk).toString();
          } catch {
            content = chunk.toString();
          }
          text += content.replace(/<([0-9A-Fa-f]+)>/g, (_, hex) =>
            Buffer.from(hex, 'hex').toString()
          );
          idx = end + 9;
        }
        return text;
      };
      const rawBrowser = extract(browserPdf);
      const rawFallback = extract(fallbackPdf);
      expect(rawBrowser).toContain('Bullet point');
      expect(rawFallback).toContain('Bullet point');
    }
  });

  test('generated PDF preserves line breaks within list items', async () => {
    const input = 'Jane Doe\n- First line\nSecond line';
    const pdf = await generatePdf(input, 'modern');
    const start = pdf.indexOf(Buffer.from('stream')) + 6;
    const nl = pdf.indexOf('\n', start) + 1;
    const end = pdf.indexOf(Buffer.from('endstream'), nl);
    const content = zlib.inflateSync(pdf.slice(nl, end)).toString();
    const matches = [...content.matchAll(/1 0 0 1 57\.536 ([0-9.]+) Tm/g)].map((m) => parseFloat(m[1]));
    expect(matches.length).toBeGreaterThan(1);
    expect(matches[1]).toBeLessThan(matches[0]);
  });

});
