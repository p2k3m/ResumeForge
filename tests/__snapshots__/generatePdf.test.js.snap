// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePdf and parsing 2025 template renders expected HTML snapshot 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Resume</title>
  <!-- 2025 template ensures whitespace and tab rendering -->
<style>@import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;700&display=swap');

:root {
  --accent: #1f3c5d;
  --bullet: #4a5568;
}

body {
  font-family: 'Inter', sans-serif;
  margin: 2.5rem;
  color: #333;
  line-height: 1.6;
  max-width: 900px;
  margin-left: auto;
  margin-right: auto;
}

header {
  text-align: center;
  margin-bottom: 2rem;
  display: flex;
  justify-content: center;
}

h1 {
  font-size: 2.8rem;
  font-weight: 700;
  margin: 0;
  color: var(--accent);
}

h2 {
  font-size: 1.5rem;
  font-weight: 700;
  margin-top: 2.5rem;
  margin-bottom: 0.75rem;
  color: var(--accent);
  border-bottom: 2px solid var(--accent);
  padding-bottom: 0.25rem;
}

main {
  display: grid;
  gap: 2rem;
}

section {
  padding: 1.25rem;
  border: 1px solid #e0e0e0;
  border-radius: 8px;
  background: #fafafa;
}

@media (min-width: 700px) {
  main {
    grid-template-columns: repeat(2, 1fr);
  }
}

ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

li {
  /* preserve formatting for multi-line bullets */
  margin-bottom: 0.5rem;
  margin-left: 1.2em;
  white-space: pre-wrap;
  line-height: 1.5;
  font-weight: 400;
}

.bullet, .edu-bullet {
  display: inline-block;
  width: 1.2em;
  margin-left: -1.2em;
  color: var(--bullet);
}

.tab {
  display: inline-block;
  width: 1.5em;
}

.tab + p {
  margin-top: 0;
}

strong {
  font-weight: 700;
}

</style></head>
<body>
  <header class="header">
    <h1>Jane Doe</h1>
  </header>
  <main class="content">
    
    <section class="section">
      <h2>Skills</h2>
      {{#if items}}
        <ul>
          {{#each items}}
            <li>{{{this}}}</li>
          
    <section class="section">
      <h2>Work Experience</h2>
      {{#if items}}
        <ul>
          {{#each items}}
            <li>{{{this}}}</li>
          
    <section class="section">
      <h2>Education</h2>
      {{#if items}}
        <ul>
          {{#each items}}
            <li>{{{this}}}</li>
          
        </ul>
      {{/if}}
    </section>
    {{/each}}
  </main>
</body>
</html>
"
`;
